abstract_domains/numerical/octagon.py

```diff
@@ -0,0 +1,578 @@
+from copy import deepcopy
+from enum import Enum
+from functools import reduce
+
+from abstract_domains.lattice import BottomMixin
+from abstract_domains.numerical.dbm import IntegerCDBM
+from abstract_domains.numerical.interval import IntervalLattice, IntervalDomain
+from abstract_domains.numerical.numerical import NumericalMixin
+from abstract_domains.state import State
+from core.expressions import *
+from typing import List, Set, Tuple, Union
+from math import inf, isinf
+
+from abstract_domains.numerical.linear_forms import SingleVarLinearForm, LinearForm, InvalidFormError
+
+from core.expressions_tools import ExpressionVisitor, ExpressionTransformer, \
+    make_condition_not_free, simplify
+
+Sign = UnaryArithmeticOperation.Operator
+PLUS = Sign.Add
+MINUS = Sign.Sub
+
+
+def _index_shift(sign: Sign):
+    return 1 if sign == MINUS else 0
+
+
+class OctagonLattice(BottomMixin, NumericalMixin):
+    def __init__(self, variables: List[VariableIdentifier]):
+        """Create an Octagon Lattice for given variables.
+        
+        :param variables: list of program variables
+        """
+        self._variables = variables
+        self._var_to_index = {}
+        self._index_to_var = {}
+        index = 0
+        for var in self.variables:
```

Comment:
It would be nice to have some (line) comments to explain what is going on here. For instance, with these two lines:
```
self._index_to_var[index] = var
self._index_to_var[index + 1] = var
```